java-checks/src/main/java/org/sonar/java/checks/security/AndroidPersistentUniqueIdentifierCheck.java

```diff
@@ -0,0 +1,125 @@
+    MethodMatchers.create()
+      .ofTypes("android.provider.Settings$Secure")
+      .names("getString")
+      .addParametersMatcher("android.content.ContentResolver", "java.lang.String")
+      .build();
+
+  private static final Map<MethodMatchers, String> matchers = Map.of(
+    MethodMatchers.create()
+      .ofTypes("android.bluetooth.BluetoothAdapter")
+      .names("getAddress")
+      .withAnyParameters()
+      .build(),
+    HARDWARE_ID_MESSAGE,
+    MethodMatchers.create()
+      .ofSubTypes("android.net.wifi.WifiInfo")
+      .names("getMacAddress")
+      .withAnyParameters()
+      .build(),
+    HARDWARE_ID_MESSAGE,
+    MethodMatchers.create()
+      .ofSubTypes("android.telephony.TelephonyManager")
+      .names("getSimSerialNumber", "getDeviceId", "getImei", "getMeid")
+      .withAnyParameters()
+      .build(),
+    HARDWARE_ID_MESSAGE,
+    MethodMatchers.create()
+      .ofSubTypes("android.telephony.TelephonyManager")
+      .names("getLine1Number")
+      .withAnyParameters()
+      .build(),
+    PHONE_NUMBER_MESSAGE,
+    MethodMatchers.create()
+      .ofSubTypes("android.telephony.SubscriptionManager")
+      .names("getPhoneNumber")
+      .withAnyParameters()
+      .build(),
+    PHONE_NUMBER_MESSAGE,
+    MethodMatchers.create()
+      .ofSubTypes(
+        "com.google.android.gms.ads.identifier.AdvertisingIdClient$Info",
+        "androidx.ads.identifier.AdvertisingIdInfo",
+        "com.huawei.hms.ads.identifier.AdvertisingIdClient$Info"
+      )
+      .names("getId")
+      .withAnyParameters()
+      .build(),
+    ADVERTISING_ID_MESSAGE,
+    staticSettingsSecureGetStringFunMatcher,
+    NON_RESETTABLE_PERSISTENT_ID_MESSAGE
+  );
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return List.of(Tree.Kind.METHOD_INVOCATION, Tree.Kind.METHOD_REFERENCE);
+  }
+
+  @Override
+  protected MethodMatchers getMethodInvocationMatchers() {
+    return MethodMatchers.or(matchers.keySet().toArray(MethodMatchers[]::new));
+  }
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree mit) {
+    matchers.keySet().stream()
+      .filter(matcher -> matcher.matches(mit))
+      .findFirst()
+      .filter(matcher ->
+        matcher != staticSettingsSecureGetStringFunMatcher ||
+          mit.arguments().get(1).asConstant().map("android_id"::equals).orElse(false))
+      .ifPresent(matcher -> {
+        var tree = mit.methodSelect().is(Tree.Kind.MEMBER_SELECT)
+          ? ((MemberSelectExpressionTree) mit.methodSelect()).identifier()
+          : mit.methodSelect();
+        var message = matchers.get(matcher);
+        reportIssue(tree, message);
+      });
+  }
+
+  @Override
+  protected void onMethodReferenceFound(MethodReferenceTree methodReferenceTree) {
+    matchers.keySet().stream()
+      .filter(matcher -> matcher.matches(methodReferenceTree))
+      .findFirst()
+      .filter(Predicate.not(staticSettingsSecureGetStringFunMatcher::equals))
```

Comment:
Similarly to the method above, is there a way we can combine the two filters in a single static boolean method?

Reply:
Yes, as above, I could coalesce the two filters into a single `isReferenceCandidate` method.